tls/s2n_server_cert_request.c

```diff
@@ -82,11 +82,83 @@ static int s2n_recv_client_cert_preferences(struct s2n_stuffer *in, s2n_cert_typ
+    struct s2n_cert_chain_and_key *chain;
+};
+
+struct s2n_certificate_authority_list *s2n_certificate_request_get_ca_list(struct s2n_certificate_request *request)
+{
+    return request != NULL ? &request->list : NULL;
+}
+
+int s2n_certificate_request_set_certificate(struct s2n_certificate_request *request, struct s2n_cert_chain_and_key *chain)
+{
+    POSIX_ENSURE(request, S2N_ERR_INVALID_ARGUMENT);
+    POSIX_ENSURE(chain, S2N_ERR_INVALID_ARGUMENT);
+    request->chain = chain;
+    return S2N_SUCCESS;
+}
+
+bool s2n_certificate_authority_list_has_next(struct s2n_certificate_authority_list *list)
+{
+    return list != NULL && s2n_stuffer_data_available(&list->iterator) > 0;
+}
+
+S2N_RESULT s2n_certificate_authority_list_read_next(struct s2n_certificate_authority_list *list, uint8_t **name, uint16_t *length)
+{
+    RESULT_ENSURE_REF(list);
+    RESULT_ENSURE_MUT(length);
+    RESULT_ENSURE_MUT(name);
+
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&list->iterator, length));
+    RESULT_ENSURE_GT(*length, 0);
+
+    *name = s2n_stuffer_raw_read(&list->iterator, *length);
+    RESULT_ENSURE_REF(*name);
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_certificate_authority_list_next(struct s2n_certificate_authority_list *list, uint8_t **name, uint16_t *length)
+{
+    POSIX_ENSURE(list, S2N_ERR_INVALID_ARGUMENT);
+    POSIX_ENSURE(name, S2N_ERR_INVALID_ARGUMENT);
+    POSIX_ENSURE(length, S2N_ERR_INVALID_ARGUMENT);
+
+    /* Ensure that if we exit due to errors we've cleared these fields */
+    *name = NULL;
+    *length = 0;
+
+    POSIX_ENSURE(s2n_certificate_authority_list_has_next(list), S2N_ERR_INVALID_ARGUMENT);
+    POSIX_ENSURE(s2n_result_is_ok(s2n_certificate_authority_list_read_next(list, name, length)), S2N_ERR_BAD_MESSAGE);
+    return S2N_SUCCESS;
+}
+
+int s2n_certificate_authority_list_reread(struct s2n_certificate_authority_list *list)
+{
+    POSIX_ENSURE(list, S2N_ERR_INVALID_ARGUMENT);
+    return s2n_stuffer_reread(&list->iterator);
+}
+
 static int s2n_set_cert_chain_as_client(struct s2n_connection *conn)
 {
     if (s2n_config_get_num_default_certs(conn->config) > 0) {
-        struct s2n_cert_chain_and_key *cert = s2n_config_get_single_default_cert(conn->config);
-        POSIX_ENSURE_REF(cert);
+        struct s2n_cert_chain_and_key *cert = NULL;
+        if (conn->config->cert_request_cb) {
+            struct s2n_certificate_request request = { 0 };
+            /* Don't try to free this stuffer, it's purely for reading pre-existing blob that outlives it. */
+            POSIX_GUARD(s2n_stuffer_init_written(&request.list.iterator, &conn->cert_authorities));
+            POSIX_ENSURE(conn->config->cert_request_cb(conn, conn->config->cert_request_cb_ctx, &request) == S2N_SUCCESS, S2N_ERR_CANCELLED);
+            cert = request.chain;
+        } else {
+            cert = s2n_config_get_single_default_cert(conn->config);
+        }
+        POSIX_ENSURE(cert, S2N_ERR_NO_CERT_FOUND);
```

Comment:
nit - let's keep the existing internal error for `s2n_config_get_single_default_cert()`:
```suggestion
            POSIX_ENSURE(conn->config->cert_request_cb(conn, conn->config->cert_request_cb_ctx, &request) == S2N_SUCCESS, S2N_ERR_CANCELLED);
            cert = request.chain;
            POSIX_ENSURE(cert, S2N_ERR_NO_CERT_FOUND);
        } else {
            cert = s2n_config_get_single_default_cert(conn->config);
            POSIX_ENSURE_REF(cert);
        }
```